### ISSUE TITLE
ActiveQuery#removeDuplicatedModels should throw an exception if primary key is not defined

### ISSUE THREAD

**Issue by dryu** [yiisoft/yii2#7780]
Otherwise, it throws a very vague error which took me some time to understand:

```
2015-03-18 10:30:28 [127.0.0.1][patyua][7klq7mn1oilt0dgtl6ogc5jki1][error][yii\base\UnknownPropertyException] exception 'yii\base\UnknownPropertyException' with message 'Getting unknown property: dryu\qtool\db\crmspx\IncidentNote::' in C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Component.php:143
Stack trace:
#0 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\BaseActiveRecord.php(247): yii\base\Component->__get('')
#1 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Model.php(926): yii\db\BaseActiveRecord->__get('')
#2 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\ActiveQuery.php(263): yii\base\Model->offsetGet(false)
#3 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\ActiveQuery.php(220): yii\db\ActiveQuery->removeDuplicatedModels(Array)
#4 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\Query.php(207): yii\db\ActiveQuery->populate(Array)
#5 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\ActiveQuery.php(130): yii\db\Query->all(NULL)
#6 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\ActiveRelationTrait.php(157): yii\db\ActiveQuery->all()
#7 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\db\BaseActiveRecord.php(249): yii\db\ActiveQuery->findFor('notesEx', Object(dryu\qtool\db\mssql\Incident))
#8 C:\inetpub\sites\qtyii2\vendor\dryu\qtool\sr\IncidentEventsGenerator.php(34): yii\db\BaseActiveRecord->__get('notesEx')
#9 C:\inetpub\sites\qtyii2\vendor\dryu\qtool\sr\IncidentEventsGenerator.php(30): dryu\qtool\sr\IncidentEventsGenerator->generateNoteEvents()
#10 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Object.php(107): dryu\qtool\sr\IncidentEventsGenerator->init()
#11 C:\inetpub\sites\qtyii2\frontend\views\view\sr\events.php(17): yii\base\Object->__construct(Array)
#12 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\View.php(325): require('C:\\inetpub\\site...')
#13 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\View.php(247): yii\base\View->renderPhpFile('C:\\inetpub\\site...', Array)
#14 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\View.php(149): yii\base\View->renderFile('C:\\inetpub\\site...', Array, Object(frontend\controllers\ViewController))
#15 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Controller.php(398): yii\base\View->render('sr/events', Array, Object(frontend\controllers\ViewController))
#16 C:\inetpub\sites\qtyii2\frontend\controllers\ViewController.php(86): yii\base\Controller->renderPartial('sr/events', Array)
#17 [internal function]: frontend\controllers\ViewController->actionSr('67422302')
#18 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\InlineAction.php(55): call_user_func_array(Array, Array)
#19 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Controller.php(151): yii\base\InlineAction->runWithParams(Array)
#20 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Module.php(455): yii\base\Controller->runAction('sr', Array)
#21 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\web\Application.php(83): yii\base\Module->runAction('view/sr', Array)
#22 C:\inetpub\sites\qtyii2\vendor\yiisoft\yii2\base\Application.php(375): yii\web\Application->handleRequest(Object(yii\web\Request))
#23 C:\inetpub\sites\qtyii2\frontend\web\index.php(17): yii\base\Application->run()
#24 {main}
```

Yii 2.0.2

**Comment by lynicidn** [comment]
need more info - sample code how reproduce it

**Comment by dryu** [comment]
It is all evident from the code:

```php
    /**
     * Removes duplicated models by checking their primary key values.
     * This method is mainly called when a join query is performed, which may cause duplicated rows being returned.
     * @param array $models the models to be checked
     * @return array the distinctive models
     */
    private function removeDuplicatedModels($models)
    {
        $hash = [];
        /* @var $class ActiveRecord */
        $class = $this->modelClass;
        $pks = $class::primaryKey();

        if (count($pks) > 1) {
            foreach ($models as $i => $model) {
                $key = [];
                foreach ($pks as $pk) {
                    $key[] = $model[$pk];
                }
                $key = serialize($key);
                if (isset($hash[$key])) {
                    unset($models[$i]);
                } else {
                    $hash[$key] = true;
                }
            }
        } else {
            $pk = reset($pks);
            foreach ($models as $i => $model) {
                $key = $model[$pk];
                if (isset($hash[$key])) {
                    unset($models[$i]);
                } elseif ($key !== null) {
                    $hash[$key] = true;
                }
            }
        }

        return array_values($models);
    }
```

This methods does not expect that the table might not have primary keys in it.

**Comment by lynicidn** [comment]
>unset($models[$i]);
what???

**Comment by lynicidn** [comment]
if table not have pk why u use $class::primaryKey()?

**Comment by samdark** [comment]
@lynicidn https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQuery.php#L220 It's valid request. The error could suggest adding a primary key...

**Comment by samdark** [comment]
@yiisoft/core-developers what do you think about making additional check there and throwing meaningful error?

**Comment by lynicidn** [comment]
@samdark yeah it in core ...

**Comment by qiangxue** [comment]
There are several similar places which assume the AR class has non-empty primary keys.

**Comment by cebe** [comment]
AR also does not work correctly without one. If it does not introduce too much overhead we could add it.

**Comment by klimov-paul** [comment]
Issue resolved by commit 68351e7
